src/residual_error_estimators.cc

```diff
@@ -0,0 +1,337 @@
+        const Function<spacedim>*                     sigma,
+        const Function<spacedim>*                     a
+      ) {
+      Assert(dim <= spacedim, ExcNotImplemented());
+      Assert(dim >= 2 && dim <= 3, ExcNotImplemented());
+      Assert(spacedim >= 2 && spacedim <= 3, ExcNotImplemented());
+
+      // Loop over all cells
+      unsigned int counter = 0;
+      for (const auto &cell : ts_tria->active_cell_iterators())
+      {
+        residuals(counter++) = estimate_one_cell(ts_tria, cell,
+                                                 n_gauss_points,
+                                                 rhs_fcn, solution,
+                                                 neumann_bc,
+                                                 sigma, a);
+      }
+    } // estimate
+
+      template<int dim, int spacedim>
+      double Poisson<dim, spacedim>::estimate_one_cell(
+          const TS_TriangulationBase<dim, spacedim>*  ts_tria,
+          const active_cell_iterator                  cell,
+          const std::vector< unsigned int >&          n_gauss_points,
+          const Function<spacedim>*                   rhs_fcn,
+          const Vector<double>&                       solution,
+          const std::map< 
+                        types::boundary_id,
+                  const Function<spacedim>* 
+                >&                                    neumann_bc,
+          const Function<spacedim>*                   sigma,
+          const Function<spacedim>*                   a
+      ) {
+
+        double face_residuals = 0.;
+        double cell_residual = 0.;
+        if (!neumann_bc.empty())
+          face_residuals = estimate_face_residuals(ts_tria, cell,
+                                              n_gauss_points, 
+                                              solution, neumann_bc);
+        // check, if sigma is NULL
+        bool sigma_const = false;
+        try
+        {
+          if (sigma != NULL)
+            sigma->gradient(Point<spacedim>());
+        }
+        catch (...)
+        {
+          sigma_const = true;
+        }
+        if (sigma_const == true) {
+          cell_residual = estimate_cell_residual_sigma_const(ts_tria, cell,
+                                                      n_gauss_points,
+                                                      rhs_fcn, solution,
+                                                      a);
+        }
+        else {
+          cell_residual = estimate_cell_residual(ts_tria, cell,
+                                          n_gauss_points,
+                                          rhs_fcn, solution,
+                                          sigma, a);
+        }
+        return std::sqrt(face_residuals + cell_residual); /*needs adjustment, but what exactly?*/
+      } // estimate_one_cell
+
+      template<int dim, int spacedim>
+      double Poisson<dim, spacedim>::estimate_face_residuals(
```

Comment:
It seems like you are calculating the face residuals only from the current cell `cell`. Is that correct? 

For C0 edges, we also need the face residuals from the neighboring side to get the jumps in the inner-part of the domain correctly: 
$$\eta(\mathrm{F}) = \frac12 \left(\frac{\partial u_h}{\partial n} - \frac{\partial u_h}{\partial n'} \right),$$
where $`n`$ is normal at $`\mathrm{F}`$ on cell $`\mathrm{Q}`$ and $`n'`$ is normal at $`\mathrm{F}`$ on cell $`\mathrm{Q}'`$ with $`\mathrm{Q}\cap\mathrm{Q}'=\mathrm{F}`$.

Reply:
This could be achieved by iterating the neighbors of the cells.